text/0000-offline-buildpackages.md

```diff
@@ -0,0 +1,479 @@
+
+  [[assets.exclude-assets]]
+    sha256 = "some-sha256"
+
+  [[assets.asset-override]]
+    uri = "preferred/asset/location"
+    sha256 = "asset-sha256"
+```
+
+In the above example we can use the `asset-caches` array to specify multiple uris to a number of asset caches. The assets from each of these will be filtered according to the `exclude-buildpack-assets`, `exclude-assets` and `asset-override` fields. assets that pass through these filterings are then added to the resulting builder.
+
+To ease adding assets to a builder we propose adding the `pack create-builder --asset-cache` flag which may specify the uri of an asset cache. This flag may be used multiple times to add multiple asset caches. All assets in the specified caches will then be added to the builder.
+
+Finally there should be the option to include all assets from specified buildpackages using the `--offline` flag (the resulting builders will be very large).
+
+## Using Asset Caches
+
+Asset caches can be added to a build by two mechanisms
+1) specify an asset cache(s) using the `--asset-cache` flag(s) in a `pack build <...>`.
+2) Assets cache layers can be added to a builder image during its creation. These assets will then be available to all builds that use this builder.
+
+
+When asset caches are added to a build/builder we need to verify the following:
+  - The asset is not excluded through a `asset-cache.toml` file passed to `pack build`.
+  - If two assets provide the same `/cnb/assets/<asset-digest>` these two files must have identical contents. Decisions about rewriting these layers to optimize space are left to the platform.
+
+When creating a builder with asset caches:
+  - If any asset layer is a superset of another, only the superset layer is included in the builder.
+  - builders have a `io.buildpacks.buildpack.assets` Label containing entries for every asset included in the builder.
+
+
+### `lifecycle` changes
+Builders should now provide a `CNB_ASSETS` environment variable to the `build` phase`. This provides a standard variable buildpacks may use when looking up assets. The value of `CNB_ASSETS` will default to `/cnb/assets`.
+
+### Platform container setup
+
+When performing a build. The platform should apply asset layer changesets to the build container. As a result there will be a filesystem subtree rooted at `/cnb/assets/` E.g.
+
+```
+cnb
+ └── assets
+     ├── <java11-asset-sha256>
+     ├── <java13-asset-sha256>
+     └── <java15-asset-sha256>
+```
+
+Buildpacks should then be able to quickly check for available assets using a `sha256` of the asset they wish to use in tandem with the `CNB_ASSETS` environment variable. The `sha256` value used to perform this lookup should be pulled from the relevant `[[asset]]` entry in `buildpack.toml`
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Why should we *not* do this?
+
+- Increases complexity for users/buildpack authors. This adds another artifact to manage and new code paths for checking for vendored assets.
```

Comment:
It is worth noting that this will add considerable complexity for kpack users who now need to provide a buildpackage image and a asset image when looking to create an "offline" builder. Would it make sense in a future rfc to explore a buildpackage format that can include assets?

Reply:
I'm hopeful the ability to construct an asset-cache from a buildpackage, and the ability to build offline builders from online buildpackages will ease this pain point.  But if not It might be worthwhile to think about merging buildpackages & assets.